password/src/Data/Password/Validate.hs

```diff
@@ -0,0 +1,276 @@
+import Data.Semigroup ((<>))
+#endif
+import Data.Text (Text)
+import qualified Data.Text as T
+
+-- $setup
+-- >>> :set -XOverloadedStrings
+--
+-- Import needed libraries.
+--
+-- >>> import Data.Password
+
+-- | Set of policies used to validate 'Password'
+data PasswordPolicy = PasswordPolicy
+    { minimumLength  :: !Int
+    -- ^ Required password minimum length
+    , maximumLength  :: !Int
+    -- ^ Required password maximum length
+    , uppercaseChars :: !(Maybe Int)
+    -- ^ Required number of upper-case characters
+    , lowercaseChars :: !(Maybe Int)
+    -- ^ Required number of lower-case characters
+    , specialChars   :: !(Maybe Int)
+    -- ^ Required number of special characters
+    , digitChars     :: !(Maybe Int)
+    -- ^ Required number of ASCII-digit characters
+    } deriving (Eq, Ord, Show)
+
+-- | Default value for the 'PasswordPolicy'
+defaultPasswordPolicy :: PasswordPolicy
+defaultPasswordPolicy = PasswordPolicy
+  { minimumLength = 8,
+    maximumLength = 32,
+    uppercaseChars = Just 1,
+    lowercaseChars = Just 1,
+    specialChars = Nothing,
+    digitChars = Just 1
+  }
+
+-- | Predicate which defines the characters that can be used for a password.
+newtype CharSetPredicate = CharSetPredicate
+  { getCharSetPredicate :: Char -> Bool
+  }
+
+
+-- | Default character sets consist of uppercase, lowercase letters, numbers,
+-- and special characters
+defaultCharSetPredicate :: CharSetPredicate
+defaultCharSetPredicate =  CharSetPredicate $ \c -> ord c >= 32 && ord c <= 126
+
+-- | Check if given 'Char' is special character
+isSpecial :: Char -> Bool
+isSpecial = \c ->
+    isDefault c && not (or [isAsciiUpper c, isAsciiLower c, isDigit c])
+  where
+    CharSetPredicate isDefault = defaultCharSetPredicate
+
+-- | Character Category
+data CharacterCategory
+  = Uppercase
+  -- ^ Uppercase letters
+  | Lowercase
+  -- ^ Lowercase letters
+  | Special
+  -- ^ Special characters
+  | Digit
+  -- ^ ASCII digits
+  deriving (Eq, Ord, Show)
+
+-- | Convert 'CharacterCategory' into associated predicate function
+categoryToPredicate :: CharacterCategory -> (Char -> Bool)
+categoryToPredicate = \case
+  Uppercase -> isAsciiUpper
+  Lowercase -> isAsciiLower
+  Special -> isSpecial
+  Digit -> isDigit
+
+-- | Possible reason of password being invalid
+data InvalidReason
+  = PasswordTooShort !Int !Int
+  -- ^ Length of 'Password' is too short.
+  --
+  -- Expected at least 'Int' characters but the actual length is 'Int'
+  | PasswordTooLong !Int !Int
+  -- ^ Length of 'Password' is too long.
+  --
+  -- Expected at maximum of 'Int' characters, but the actual length is 'Int'
+  | NotEnoughReqChars CharacterCategory !Int !Int
+  -- ^ 'Password' does not contain required number of characters.
+  --
+  -- Expected at least 'Int' characters of 'CharacterCategory' but the password only
+  -- contains 'Int'
+  | InvalidCharacters !Text
+  -- ^ 'Password' contains characters that cannot be used
+  | InvalidLength !Int !Int
```

Comment:
A good idea might be to put all policy-related `InvalidReason`s in a separate type (e.g. `InvalidPolicyReason`) which can be put in:
```hs
data InvalidReason =
  ...
  | InvalidPolicy [InvalidPolicyReason]
```
so that it's easier to check if the policy is bad, rather than the password.
This would also make it easier to print nice user-facing messages for the password reasons and give a generic "oops, something is wrong on our end" reason.

Reply:
Personally, whether this would be a good idea is very subtle. 
The fact that this caused the test code to be considerably more cumbersome make me thinks this is not a simple improvement.
I'll push the commit so you can take a look at it, but I'd prefer this commit to be reverted and moved to separate PR.

Reply:
Sure, making extra branches is always a good thing. Better to have alternatives if anything is up for debate.

I do also have strong feelings about the tests, though. I don't believe the property tests that use `ValidPassword(Policy)` and `InvalidPassword(Policy)` like this are a good way to test the `Data.Password.Validate` module, since you're just implementing the same logic twice, but with more boilerplate.
Most of the tests IMHO can be done easily with unit tests and are probably more readable too.

Reply:
@Vlix 
I'm not going to rewrite the tests unless there's a definite advantage over it or you think the code is un-maintainable (In that case, the response of the [comment](https://github.com/cdepillabout/password/pull/21#issuecomment-630704197) should've been "Let's start with unit tests")

Unit testing and property testing each have their advantages and disadvantages.

This module contains a lot of conditional expressions, and functions that return several different values. Therefore, the challenge was how to increase the test coverage.

If it is a unit test, it can be implemented with highly readable codes. On the other hand, a huge amount of test cases are required to increase coverage, and human error is prone to occur when necessary test cases are overlooked. Also, most of the code will definitely be duplicate codes.

With property testing, on the other hand, you can get close to 100% coverage while keeping the number of tests low. In fact, during the implementation, I was able to find many bugs that would have been overlooked if not for the property testing. On the other hand, as you say, some codes become boilerplate, and troublesome to maintain.

In other words, both tests have their own strengths and weaknesses, and it was up to me to decide which one to take. Since this package is about security, I have decided that property testing is appropriate.

I also [raised the choice of testing method as a concern](https://github.com/cdepillabout/password/pull/21#issuecomment-630704197) when I initially worked on this issue. The answer at the time was, "Yup, you have free reign on how to implement all the validation stuff here". That said, it's an irresponsible statement to say that it's "probably" better to rewrite everything into unit tests three weeks after the comment.

It's also mentally painful to have to rewrite code which I've put my effort into it just because the unit tests are "probably" better.

Reply:
@cdepillabout gave free reign, yeah, my [answer to your question](https://github.com/cdepillabout/password/pull/21#issuecomment-633055194) was a bit more subtle.
And it's not like you have to ditch all your test code. I might have come off a bit too harsh in my previous comment. The property tests will still test a lot of things, I just feel less comfortable with depending on JUST these property tests in this case. If you feel it's enough, that's fine, I'll just add some more when your done, just to be doubly sure. And any critique is always welcome. Property vs. Unit tests have pros and cons either way.

Like I said, I appreciate all the effort and I have the nasty habit of commenting on almost everything. I'm trying to keep it to a mininum, so please forgive me.